tests/subsys/bluetooth/mesh/scheduler_model/message_validity/src/main.c

```diff
@@ -0,0 +1,477 @@
+#include <scheduler_internal.h>
+#include <model_utils.h>
+#include <time_util.h>
+#include <sched_test.h>
+
+#define TIMEOUT 5
+
+static uint8_t test_idx;
+static struct tm start_tm;
+struct bt_mesh_schedule_entry valid_entry;
+
+struct bt_mesh_msg_ctx test_ctx = {
+	.addr = 1,
+};
+
+K_SEM_DEFINE(action_set_cb_called, 0, 1);
+K_SEM_DEFINE(action_status_rcvd, 0, 1);
+K_SEM_DEFINE(status_rcvd, 0, 1);
+K_SEM_DEFINE(status_send, 0, 2);
+
+static struct bt_mesh_time_srv time_srv = BT_MESH_TIME_SRV_INIT(NULL);
+
+static void onoff_set(struct bt_mesh_onoff_srv *srv, struct bt_mesh_msg_ctx *ctx,
+		      const struct bt_mesh_onoff_set *set, struct bt_mesh_onoff_status *rsp)
+{
+}
+
+static void onoff_get(struct bt_mesh_onoff_srv *srv, struct bt_mesh_msg_ctx *ctx,
+		      struct bt_mesh_onoff_status *rsp)
+{
+}
+
+struct bt_mesh_onoff_srv_handlers onoff_srv_handlers = {
+	.set = onoff_set,
+	.get = onoff_get,
+};
+
+struct bt_mesh_onoff_srv onoff_srv = BT_MESH_ONOFF_SRV_INIT(&onoff_srv_handlers);
+
+static void srv_action_set_cb(struct bt_mesh_scheduler_srv *srv, struct bt_mesh_msg_ctx *ctx,
+			      uint8_t test_idx, struct bt_mesh_schedule_entry *entry)
+{
+	k_sem_give(&action_set_cb_called);
+}
+
+static struct bt_mesh_scheduler_srv sched_srv =
+	BT_MESH_SCHEDULER_SRV_INIT(&srv_action_set_cb, &time_srv);
+
+static void cli_action_status_handler(struct bt_mesh_scheduler_cli *cli,
+				      struct bt_mesh_msg_ctx *ctx, uint8_t test_idx,
+				      const struct bt_mesh_schedule_entry *action)
+{
+	k_sem_give(&action_status_rcvd);
+}
+
+static void cli_status_handler(struct bt_mesh_scheduler_cli *cli, struct bt_mesh_msg_ctx *ctx,
+			       uint16_t schedules)
+{
+	k_sem_give(&status_rcvd);
+}
+
+static struct bt_mesh_scheduler_cli sched_cli = {
+	.action_status_handler = &cli_action_status_handler,
+	.status_handler = &cli_status_handler,
+};
+
+/** Mocking mesh composition data
+ *
+ * The mocked composition data consists of 1 element with
+ * 1 Scheduler server, 1 Scheduler client, and 1 Generic onoff server.
+ *
+ * |      1     |
+ * |:----------:|
+ * | Sched. Srv |
+ * | Sched. Cli |
+ * | Onoff Srv  |
+ *
+ * API doc: If the Scheduler server is registered on the element then the Scene server shouldn't.
+ * The Scheduler model already includes the Scene model.
+ */
+static struct bt_mesh_model mock_models_elem[3] = {
+	{
+		.user_data = &sched_srv,
+		.id = BT_MESH_MODEL_ID_SCHEDULER_SRV,
+	},
+	{
+		.user_data = &sched_cli,
+		.id = BT_MESH_MODEL_ID_SCHEDULER_CLI,
+	},
+	{
+		.user_data = &onoff_srv,
+		.id = BT_MESH_MODEL_ID_GEN_ONOFF_SRV,
+	},
+};
+
+static struct bt_mesh_elem mock_elems[1] = {
+	{ .addr = 1, .model_count = ARRAY_SIZE(mock_models_elem), .models = mock_models_elem },
+};
+
+struct mock_comp_data {
+	size_t elem_count;
+	struct bt_mesh_elem *elem;
+} mock_comp = { .elem_count = ARRAY_SIZE(mock_elems), .elem = mock_elems };
+
+static struct bt_mesh_model *sched_mod1_elem = &mock_models_elem[0];
```

Comment:
A bit confusing that these are called `..._elem` when they are model pointers, not elem pointers. You could e.g. call them `srv_mod` and `cli_mod`, might make things a bit more readable.